File: src/main/java/minic/uilocal/editor/MiniCRealtimeAnalyzer.java
package minic.uilocal;

import javafx.application.Platform;
import minic.uiapi.MiniCRealtimeAnalysisApi;
import minic.uiapi.UiRealtimeAnalysisDto;

import java.util.Objects;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * 后台实时 lexer/parser/semantic 分析器。
 */
public final class MiniCRealtimeAnalyzer implements AutoCloseable {
    private final BlockingQueue<Request> queue = new LinkedBlockingQueue<>();
    private final ResultSink resultSink;
    private final MiniCRealtimeAnalysisApi api = new MiniCRealtimeAnalysisApi();
    private volatile boolean running = true;
    private Thread worker;
    private long nextVersion;

    /**
     * 创建实时分析器。
     *
     * @param resultSink 结果回调
     */
    public MiniCRealtimeAnalyzer(ResultSink resultSink) {
        this.resultSink = Objects.requireNonNull(resultSink, "resultSink");
    }

    /**
     * 提交一次编辑输入。
     *
     * @param sourceName 源码名称
     * @param sourceText 源码文本
     */
    public void submit(String sourceName, String sourceText) {
        ensureStarted();
        queue.offer(new Request(sourceName, sourceText, ++nextVersion));
    }

    @Override
    public void close() {
        running = false;
        if (worker != null) {
            worker.interrupt();
        }
    }

    private synchronized void ensureStarted() {
        if (worker != null) {
            return;
        }
        worker = new Thread(this::runLoop, "minic-realtime-analyzer");
        worker.setDaemon(true);
        worker.start();
    }

    private void runLoop() {
        while (running) {
            try {
                Request request = queue.take();
                Request latest = drainLatest(request);
                UiRealtimeAnalysisDto result = api.analyze(latest.sourceName(), latest.sourceText(), latest.version());
                Platform.runLater(() -> resultSink.accept(result));
            } catch (InterruptedException exception) {
                if (!running) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }
    }

    private Request drainLatest(Request request) {
        Request latest = request;
        Request next;
        while ((next = queue.poll()) != null) {
            latest = next;
        }
        return latest;
    }

    /**
     * 实时分析结果回调。
     */
    @FunctionalInterface
    public interface ResultSink {
        /**
         * 接收分析结果。
         *
         * @param result 分析结果
         */
        void accept(UiRealtimeAnalysisDto result);
    }

    private record Request(String sourceName, String sourceText, long version) {
        private Request {
            Objects.requireNonNull(sourceName, "sourceName");
            Objects.requireNonNull(sourceText, "sourceText");
        }
    }
}

File: uiweb/src/editor/MiniCRealtimeAnalyzer.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiRealtimeAnalysisDto } from "../translation/uiapi";
import { requireValue } from "../translation/uiTypes";
import type { MiniCRealtimeAnalysisApiAdapter } from "../workbench/MiniCWorkbenchViewModel";

export const miniCRealtimeAnalyzerMirror = {
  "javaPath": "src/main/java/minic/uilocal/editor/MiniCRealtimeAnalyzer.java",
  "webPath": "uiweb/src/editor/MiniCRealtimeAnalyzer.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCRealtimeAnalyzer",
  "kind": "class",
  "imports": [
    "java.util.Objects",
    "java.util.concurrent.BlockingQueue",
    "java.util.concurrent.LinkedBlockingQueue",
    "javafx.application.Platform",
    "minic.uiapi.MiniCRealtimeAnalysisApi",
    "minic.uiapi.UiRealtimeAnalysisDto"
  ],
  "fields": [
    {
      "name": "api",
      "signature": "private final MiniCRealtimeAnalysisApi api="
    },
    {
      "name": "nextVersion",
      "signature": "private long nextVersion"
    },
    {
      "name": "queue",
      "signature": "private final BlockingQueue<Request>queue="
    },
    {
      "name": "resultSink",
      "signature": "private final ResultSink resultSink"
    },
    {
      "name": "running",
      "signature": "private volatile boolean running="
    },
    {
      "name": "worker",
      "signature": "private Thread worker"
    }
  ],
  "methods": [
    {
      "name": "close",
      "signature": "close()"
    },
    {
      "name": "drainLatest",
      "signature": "drainLatest(Request request)"
    },
    {
      "name": "ensureStarted",
      "signature": "ensureStarted()"
    },
    {
      "name": "Request",
      "signature": "Request(String sourceName,String sourceText,long version)"
    },
    {
      "name": "runLoop",
      "signature": "runLoop()"
    },
    {
      "name": "submit",
      "signature": "submit(String sourceName,String sourceText)"
    }
  ]
} as const satisfies JavaMirrorFile;

export type MiniCRealtimeResultSink = (result: UiRealtimeAnalysisDto) => void;

interface Request {
  readonly sourceName: string;
  readonly sourceText: string;
  readonly version: number;
}

export class MiniCRealtimeAnalyzer {
  static readonly mirror = miniCRealtimeAnalyzerMirror;

  readonly mirror = miniCRealtimeAnalyzerMirror;

  private running = true;

  private nextVersion = 0;

  private pendingRequest: Request | null = null;

  private scheduledHandle: number | null = null;

  constructor(
    private readonly api: MiniCRealtimeAnalysisApiAdapter,
    private readonly resultSink: MiniCRealtimeResultSink,
  ) {
    requireValue(api, "api");
    requireValue(resultSink, "resultSink");
  }

  submit(sourceName: string, sourceText: string): void {
    if (!this.running) {
      return;
    }
    this.pendingRequest = {
      sourceName: requireValue(sourceName, "sourceName"),
      sourceText: requireValue(sourceText, "sourceText"),
      version: this.nextVersion + 1,
    };
    this.nextVersion += 1;
    this.ensureStarted();
  }

  close(): void {
    this.running = false;
    if (this.scheduledHandle !== null) {
      window.clearTimeout(this.scheduledHandle);
      this.scheduledHandle = null;
    }
    this.pendingRequest = null;
  }

  ensureStarted(): void {
    if (this.scheduledHandle !== null) {
      return;
    }
    this.scheduledHandle = window.setTimeout(() => {
      void this.runLoop().catch(() => {
        this.scheduledHandle = null;
      });
    }, 0);
  }

  async runLoop(): Promise<void> {
    this.scheduledHandle = null;
    if (!this.running || this.pendingRequest === null) {
      return;
    }
    const request = this.drainLatest(this.pendingRequest);
    this.pendingRequest = null;
    try {
      const result = await this.api.analyze(request.sourceName, request.sourceText, request.version);
      if (this.running && result.version === request.version) {
        this.resultSink(result);
      }
    } catch {
      // Realtime diagnostics are opportunistic; the workbench ViewModel owns user-facing failures.
    }
    if (this.pendingRequest !== null) {
      this.ensureStarted();
    }
  }

  drainLatest(request: Request): Request {
    return this.pendingRequest ?? request;
  }

  summary(): string {
    return `MiniCRealtimeAnalyzer: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

export default MiniCRealtimeAnalyzer;
